pkg/operator/configobservation/clustername/observeinfraid.go

```diff
@@ -0,0 +1,40 @@
+package clustername
+
+import (
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+
+	configv1 "github.com/openshift/api/config/v1"
+	"github.com/openshift/cluster-kube-controller-manager-operator/pkg/operator/configobservation"
+	"github.com/openshift/library-go/pkg/operator/configobserver"
+	"github.com/openshift/library-go/pkg/operator/events"
+)
+
+// ObserveInfraID fills in the cluster-name extended argument for the controller-manager with the cluster's infra ID
+func ObserveInfraID(genericListers configobserver.Listers, recorder events.Recorder, existingConfig map[string]interface{}) (map[string]interface{}, []error) {
+	listers := genericListers.(configobservation.Listers)
+	errs := []error{}
+	clusterNamePath := []string{"extendedArguments", "cluster-name"}
+	previouslyObservedConfig := map[string]interface{}{}
+
+	if currentClusterName, _, _ := unstructured.NestedStringSlice(existingConfig, clusterNamePath...); len(currentClusterName) > 0 {
+		if err := unstructured.SetNestedStringSlice(previouslyObservedConfig, currentClusterName, clusterNamePath...); err != nil {
+			errs = append(errs, err)
+		}
+	}
+
+	observedConfig := map[string]interface{}{}
+	infrastructure, err := listers.InfrastructureLister().Get("cluster")
+	if errors.IsNotFound(err) {
+		recorder.Warningf("ObserveInfraID", "Required infrastructures.%s/cluster not found", configv1.GroupName)
+		return observedConfig, errs
+	}
+	if err != nil {
+		return previouslyObservedConfig, errs
+	}
+	infraID := infrastructure.Status.InfrastructureName
```

Comment:
please add a comment explaining what will happen on upgrade cases.  Is this value always present?

Reply:
https://github.com/openshift/api/commit/c9b4e5b2d7850deea6719b30aaf38159871ed654 is always set. it was part of the API with 4.0